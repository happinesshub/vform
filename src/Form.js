import http from './http';
import FormErrors from './FormErrors';

export default class Form {
    /**
     * Create a new form instance.
     *
     * @param {Object} data
     * @param {Object} mergeData
     */
    constructor(data = {}, mergeData = {}) {
        Object.assign(this, data, mergeData);

        this.busy = false;
        this.successful = false;
        this.errors = new FormErrors();
    }

    /**
     * Get the form data.
     *
     * @return {Object}
     */
    getData() {
        const data = {};
        const ignore = ['busy', 'successful', 'errors', 'forms'];

        for (let key in this) {
            if (this.hasOwnProperty(key) && ignore.indexOf(key) < 0) {
                data[key] = this[key];
            }
        }

        return data;
    }

    /**
     * Start processing the form.
     */
    startProcessing() {
        this.errors.clear();
        this.busy = true;
        this.successful = false;
    }

    /**
     * Finish processing the form.
     */
    finishProcessing() {
        this.busy = false;
        this.successful = true;
    }

    /**
     * Clear the form.
     */
    clear() {
        this.errors.clear();
        this.successful = false;
    }

    /**
     * Reset the form fields.
     */
    reset() {
        this.clear();

        for (let key in this.getData()) {
            this[key] = '';
        }
    }

    /**
     * Send the from via a GET request.
     *
     * @param  {String} url
     * @return {Promise}
     */
    get(url) {
        return http.send('get', url, this);
    }

    /**
     * Send the from via a POST request.
     *
     * @param  {String} url
     * @return {Promise}
     */
    post(url) {
        return http.send('post', url, this);
    }

    /**
     * Send the from via a PATCH request.
     *
     * @param  {String} url
     * @return {Promise}
     */
    patch(url) {
        return http.send('patch', url, this);
    }

    /**
     * Set / Get http base url.
     *
     * @param  {String|null} url
     * @return {String|null}
     */
    static baseUrl(url = null) {
        if (url) {
            http.baseUrl = url;
        }

        return http.baseUrl;
    }

    /**
     * Set / Get http routes.
     *
     * @param  {String|null} url
     * @return {String|null}
     */
    static routes(routes = null) {
        if (routes) {
            http.routes = routes;
        }

        return http.routes;
    }
}
