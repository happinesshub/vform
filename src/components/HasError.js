export default {
    props: {
        form: {
            type: Object,
            required: true
        },
        field: {
            type: String,
            required: true
        }
    },

    template: `
        <span class="has-error" v-if="form.errors.has(field)">
            {{ form.errors.get(field) }}
        </span>
    `
}
