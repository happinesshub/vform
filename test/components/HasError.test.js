import Form from '@/Form'
import { mount } from '@vue/test-utils'
import HasError from '@/components/HasError'

describe('HasError', () => {
  test('render error message', () => {
    const form = new Form({ username: '' })
    form.errors.set({ username: 'Username is required' })

    const wrapper = mount(HasError, {
      propsData: { form, field: 'username' }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('doesn`t render error message', () => {
    const form = new Form({ username: '' })

    const wrapper = mount(HasError, {
      propsData: { form, field: 'username' }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
