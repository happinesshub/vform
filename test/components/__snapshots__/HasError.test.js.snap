// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HasError doesn\`t render error message 1`] = `undefined`;

exports[`HasError render error message 1`] = `<div class="help-block invalid-feedback">Username is required</div>`;
